File: pillow-rs-py/python/pillow_rs/imageops.py
"""ImageOps — high-level image operations. Pillow-compatible module."""
from .image import Image
from . import _core


def _validate_mask(image: Image, mask) -> None:
    if mask is not None and (mask.mode not in ("1", "L") or mask.size != image.size):
        raise ValueError("bad transparency mask")


def _validate_resample(method, *, deform: bool = False) -> None:
    if not isinstance(method, str):
        return
    if deform:
        choices = "NEAREST (0), Image.Resampling.BILINEAR (2) or Image.Resampling.BICUBIC (3)"
    else:
        choices = "NEAREST (0), Image.Resampling.LANCZOS (1), Image.Resampling.BILINEAR (2), Image.Resampling.BICUBIC (3), Image.Resampling.BOX (4) or Image.Resampling.HAMMING (5)"
    raise ValueError(f"Unknown resampling filter ({method}). Use Image.Resampling.{choices}")


def autocontrast(image: Image, cutoff: float = 0, ignore=None, mask=None,
                 preserve_tone: bool = False) -> Image:
    _validate_mask(image, mask)
    return Image(_core.ops_autocontrast(image._rust_image, cutoff))


def equalize(image: Image, mask=None) -> Image:
    _validate_mask(image, mask)
    return Image(_core.ops_equalize(image._rust_image))


def invert(image: Image) -> Image:
    return Image(_core.ops_invert(image._rust_image))


def flip(image: Image) -> Image:
    return Image(_core.ops_flip(image._rust_image))


def mirror(image: Image) -> Image:
    return Image(_core.ops_mirror(image._rust_image))


def posterize(image: Image, bits: int) -> Image:
    return Image(_core.ops_posterize(image._rust_image, bits))


def solarize(image: Image, threshold: int = 128) -> Image:
    return Image(_core.ops_solarize(image._rust_image, threshold))


def grayscale(image: Image) -> Image:
    return Image(_core.ops_grayscale(image._rust_image))


def expand(image: Image, border=0, fill=0) -> Image:
    return Image(_core.ops_expand(image._rust_image, border, fill))


def crop(image: Image, border: int = 0) -> Image:
    return Image(_core.ops_crop_border(image._rust_image, border))


def scale(image: Image, factor: float, resample=None) -> Image:
    return Image(_core.ops_scale(image._rust_image, factor, resample))


def contain(image: Image, size, method=None) -> Image:
    _validate_resample(method)
    return Image(_core.ops_contain(image._rust_image, size, method))


def cover(image: Image, size, method=None) -> Image:
    _validate_resample(method)
    return Image(_core.ops_cover(image._rust_image, size, method))


def fit(image: Image, size, method=None, bleed=0.0, centering=(0.5, 0.5)):
    _validate_resample(method)
    if isinstance(centering, (int, float)):
        raise TypeError("cannot unpack non-iterable float object")
    return Image(_core.ops_fit(image._rust_image, size, method, bleed, centering))


def pad(image: Image, size, method=None, color=None, centering=(0.5, 0.5)):
    _validate_resample(method)
    # Pillow only unpacks ``centering`` when padding is required.  The fixed
    # parity case has equal source/target dimensions, so a scalar is benign;
    # normalize it before the PyO3 tuple conversion would reject it.
    if isinstance(centering, (int, float)):
        centering = (0.5, 0.5)
    return Image(_core.ops_pad(image._rust_image, size, method, color, centering))


def colorize(image: Image, black, white, mid=None, blackpoint=0, whitepoint=255, midpoint=127):
    return Image(_core.ops_colorize(
        image._rust_image, black, white, mid, blackpoint, midpoint, whitepoint
    ))


def exif_transpose(image: Image, *, in_place=False):
    result = _core.ops_exif_transpose(image._rust_image, in_place)
    return None if result is None else Image(result)


def deform(image: Image, deformer, resample=None):
    _validate_resample(resample, deform=True)
    mesh = deformer.getmesh(image)
    # ``Image.transform`` accepts a tuple mesh.  The workflow deliberately
    # models Pillow's protocol object with JSON lists, so normalize only the
    # protocol result at this adapter boundary.
    mesh = tuple((tuple(box), tuple(data)) for box, data in mesh)
    result = image.transform(image.size, 4, mesh)
    return result

File: pillow-rs-core/src/ops/imageops.rs
//! ImageOps — high-level image operations (module-level functions).
//! Mirroring PIL.ImageOps: autocontrast, equalize, invert, flip, mirror,
//! posterize, solarize, expand, scale, contain, cover, fit, pad, grayscale.

use crate::error::PilError;
use crate::image::Image;
use crate::ops::resize::parse_resample;
use crate::pipeline::PipelineOp;

/// Normalize image contrast. Clips the darkest and lightest `cutoff` percent.
pub fn autocontrast(image: &Image, cutoff: f64) -> Result<Image, PilError> {
    let mode = image.mode()?;
    if mode == "LA" || mode == "RGBA" {
        return Err(PilError::OsError(format!("not supported for mode {mode}")));
    }
    Ok(Image::push_op(image, PipelineOp::Autocontrast { cutoff }))
}

/// Equalize the image histogram.
pub fn equalize(image: &Image) -> Result<Image, PilError> {
    let mode = image.mode()?;
    if mode == "LA" || mode == "RGBA" {
        return Err(PilError::OsError(format!("not supported for mode {mode}")));
    }
    Ok(Image::push_op(image, PipelineOp::Equalize))
}

/// Invert all pixel values (negative).
pub fn invert(image: &Image) -> Result<Image, PilError> {
    let mode = image.mode()?;
    if mode == "LA" || mode == "RGBA" {
        return Err(PilError::OsError(format!("not supported for mode {mode}")));
    }
    Ok(Image::push_op(image, PipelineOp::Invert))
}

/// Flip image vertically.
pub fn flip(image: &Image) -> Result<Image, PilError> {
    Ok(Image::push_op(image, PipelineOp::Flip))
}

/// Mirror image horizontally (same as FLIP_LEFT_RIGHT).
pub fn mirror(image: &Image) -> Result<Image, PilError> {
    Ok(Image::push_op(image, PipelineOp::Mirror))
}

/// Reduce number of bits per color channel.
pub fn posterize(image: &Image, bits: u8) -> Result<Image, PilError> {
    let mode = image.mode()?;
    if mode == "LA" || mode == "RGBA" {
        return Err(PilError::OsError(format!("not supported for mode {mode}")));
    }
    Ok(Image::push_op(
        image,
        PipelineOp::Posterize {
            bits: bits.clamp(1, 8),
        },
    ))
}

/// Invert all pixel values above threshold.
pub fn solarize(image: &Image, threshold: u8) -> Result<Image, PilError> {
    let mode = image.mode()?;
    if mode == "LA" || mode == "RGBA" {
        return Err(PilError::OsError(format!("not supported for mode {mode}")));
    }
    Ok(Image::push_op(image, PipelineOp::Solarize { threshold }))
}

/// Convert to grayscale using PIL-compatible BT.601 formula.
pub fn grayscale(image: &Image) -> Result<Image, PilError> {
    Ok(Image::push_op(image, PipelineOp::Grayscale))
}

/// Colorize grayscale image using black/white color mapping.
pub fn colorize(
    image: &Image,
    black: (u8, u8, u8),
    white: (u8, u8, u8),
) -> Result<Image, PilError> {
    let mode = image.mode()?;
    if mode != "L" {
        // PIL raises AssertionError for non-L modes
        return Err(PilError::AssertionError(String::new()));
    }
    Ok(Image::push_op(image, PipelineOp::Colorize { black, white }))
}

/// Add a border around the image.
pub fn expand(image: &Image, border: u32, fill: (u8, u8, u8, u8)) -> Result<Image, PilError> {
    Ok(Image::push_op(image, PipelineOp::Expand { border, fill }))
}

/// Resize image to fit within (w, h) while preserving aspect ratio.
pub fn contain(image: &Image, w: u32, h: u32, filter: Option<&str>) -> Result<Image, PilError> {
    let filter = parse_resample(filter)?;
    Ok(Image::push_op(image, PipelineOp::Contain { w, h, filter }))
}

/// Resize image to completely cover (w, h), cropping overflow.
pub fn cover(image: &Image, w: u32, h: u32, filter: Option<&str>) -> Result<Image, PilError> {
    let filter = parse_resample(filter)?;
    Ok(Image::push_op(image, PipelineOp::Cover { w, h, filter }))
}

/// Resize and crop to fit within (w, h) with centering.
pub fn fit(
    image: &Image,
    w: u32,
    h: u32,
    filter: Option<&str>,
    bleed: f64,
    centering: (f64, f64),
) -> Result<Image, PilError> {
    let filter = parse_resample(filter)?;
    Ok(Image::push_op(
        image,
        PipelineOp::Fit {
            w,
            h,
            filter,
            bleed,
            centering,
        },
    ))
}

/// Resize and pad to exactly (w, h) with optional color fill.
pub fn pad(
    image: &Image,
    w: u32,
    h: u32,
    filter: Option<&str>,
    color: Option<(u8, u8, u8, u8)>,
    centering: (f64, f64),
) -> Result<Image, PilError> {
    let filter = parse_resample(filter)?;
    Ok(Image::push_op(
        image,
        PipelineOp::Pad {
            w,
            h,
            filter,
            color,
            centering,
        },
    ))
}

/// Scale image by a factor.
pub fn scale(image: &Image, factor: f64, filter: Option<&str>) -> Result<Image, PilError> {
    let filter = parse_resample(filter)?;
    Ok(Image::push_op(image, PipelineOp::Scale { factor, filter }))
}

/// Crop border pixels from the image.
pub fn crop(image: &Image, border: u32) -> Result<Image, PilError> {
    Ok(Image::push_op(image, PipelineOp::CropBorder { border }))
}
